### PR TITLE
feat(app-platform): App Platform Permissions

### PR DESCRIPTION
**Pull request by mnoble**
Adds a set of scopes that are required (one of, technically) for each new App Platform endpoint recently introduced.

Also refactors the endpoint base classes and permission classes to be more explicit.

### Published vs Unpublished

Access to `SentryApp` resources is based on two things: your role within the Org who owns it as well as whether it is Published or not.

Published apps are available to everyone, while Unpublished apps are only accessible by those in the Org who owns it or superusers.

### `403` vs `404`

When a User makes a request to a resource they do not have access to, through Organization, these endpoints return a `404`, instead of the more semantically correct `403`.

The rationale is mitigate people from being able to determine whether a resource exists or not. Theoretically they could use that determine how many `<resource>`s we have, how many are installed per Org, etc.

### `SentryPermission`

I pulled the `access` portion of `OrganizationPermission` into a new baseclass: `SentryPermission`.

Some of the App Platform endpoints do not fall under an Organization, but still require the same `access` determining logic. So I broke that out into a separate base class and allows the consumer to explicitly invoke that function instead of it implicitly happening.

### Org Access Checking

Many of these baseclass `convert_args` functions already restrict Organizations to those the User is a part of, for many of the queries they make.

The various Permission classes also double check the user is a part of the Organization in question.

This is duplicative and makes some, technically, unnecessary queries, but I think the cost is small and it makes the code easier to follow.

### CODE REVIEW

**Comment by benvinegar** [comment]
👏 Great write-up – thanks.

File: src/sentry/api/bases/sentryapps.py [hunk 1]
```diff
@@ -1,70 +1,226 @@
...
+    # If there are no scopes listed, it implies a public endpoint.
+    if len(allowed_scopes) == 0:
+        return True
+
+    return any(request.access.has_scope(s) for s in set(allowed_scopes))
+
+
+class SentryAppsPermission(SentryPermission):
+    scope_map = {
+        'GET': (),  # Public endoint.
```
Comment (MeredithAnya) [inline src/sentry/api/bases/sentryapps.py]: hehe endoint

File: src/sentry/api/bases/sentryapps.py [hunk 2]
```diff
@@ -1,70 +1,226 @@
...
+            request,
+            self.scope_map.get(request.method),
+        )
+
+
+class SentryAppsBaseEndpoint(Endpoint):
+    permission_classes = (SentryAppsPermission, )
+
+    def convert_args(self, request, *args, **kwargs):
+        # Only the POST endpoint includes an Organization.
```
Comment (MeredithAnya) [inline src/sentry/api/bases/sentryapps.py]: Can we make the comment here maybe a little more descriptive, even though you went thru it with me I got confused again on https://github.com/getsentry/sentry/pull/10265/files?utf8=%E2%9C%93&diff=split#diff-7be9ee4c221344b6b849dc9a27929075R48.

File: src/sentry/api/bases/sentryapps.py [hunk 3]
```diff
@@ -1,70 +1,226 @@
...
+            self.scope_map.get(request.method),
+        )
+
+
+class SentryAppInstallationsBaseEndpoint(Endpoint):
+    permission_classes = (SentryAppInstallationsPermission, )
+
+    def convert_args(self, request, organization_slug, *args, **kwargs):
+        if request.user.is_superuser:
+            organizations = Organization.objects.all()
```
Comment (MeredithAnya) [inline src/sentry/api/bases/sentryapps.py]: Do we need this query? Seems like if we are only getting one org later we could just do `Organization.objects.filter(slug=organization_slug)` instead?

**Review by MeredithAnya — COMMENTED** [review]
Just some small comments

File: src/sentry/api/bases/sentryapps.py (hunk 3 above)
Comment (mnoble) [reply src/sentry/api/bases/sentryapps.py]: This will return a queryset, which then actually makes the query down below where we call `.get()`.

Comment (MeredithAnya) [reply src/sentry/api/bases/sentryapps.py]: I think `filter` also returns a queryset? https://docs.djangoproject.com/en/2.1/ref/models/querysets/#filter  
either way I guess it doesn't matter ¯\_(ツ)_/¯

**Review by MeredithAnya — APPROVED** [review]
🚢

**Review by marcoxavier1955 — COMMENTED** [review]
17760946_301511836948362_5529488916486294072_o 3

**Pull request merged**